include/aos/common/crypto.hpp

```diff
@@ -83,17 +79,184 @@ constexpr auto cSerialNumSize = AOS_CONFIG_CRYPTO_SERIAL_NUM_SIZE;
 constexpr auto cSerialNumStrLen = AOS_CONFIG_CRYPTO_SERIAL_NUM_STR_LEN;
 
 /**
- * General certificate private key type.
+ * RSA modulus size.
+ */
+constexpr auto cRSAModulusSize = AOS_CONFIG_CRYPTO_RSA_MODULUS_SIZE;
+
+/**
+ * Size of RSA public exponent.
+ */
+constexpr auto cRSAPubExponentSize = AOS_CONFIG_CRYPTO_RSA_PUB_EXPONENT_SIZE;
+
+/**
+ * ECDSA params OID size.
+ */
+constexpr auto cECDSAParamsOIDSize = AOS_CONFIG_CRYPTO_ECDSA_PARAMS_OID_SIZE;
+
+/**
+ * DER-encoded X9.62 ECPoint.
+ */
+constexpr auto cECDSAPointDERSize = AOS_CONFIG_CRYPTO_ECDSA_POINT_DER_SIZE;
+
+/**
+ * Supported key types.
+ */
+enum class KeyType { eRSA, eECDSA };
+
+/**
+ * Public key interface.
+ */
+class PublicKey {
+public:
+    /**
+     * Returns type of a public key.
+     *
+     * @return KeyType,
+     */
+    virtual KeyType GetKeyType() const = 0;
+
+    /**
+     * Tests whether current key is equal to the provided one.
+     *
+     * @param pubKey public key.
+     * @return bool.
+     */
+    virtual bool IsEqual(const PublicKey& pubKey) const = 0;
+
+    /**
+     * Destroys object instance.
+     */
+    virtual ~PublicKey() = default;
+};
+
+/**
+ * Private key interface.
  */
 class PrivateKey {
 public:
     /**
      * Returns public part of a private key.
+     *
+     * @return const PublicKey&.
+     */
+    virtual const PublicKey& GetPublic() const = 0;
+
+    /**
+     * Destroys object instance.
+     */
+    virtual ~PrivateKey() = default;
+};
+
+/**
+ * RSA public key.
+ */
+class RSAPublicKey : public PublicKey {
+public:
+    /**
+     * Constructs object instance.
+     *
+     * @param n modulus.
+     * @param e public exponent.
+     */
+    RSAPublicKey(const Array<uint8_t>& n, const Array<uint8_t>& e);
+
+    /**
+     * Returns type of a public key.
+     *
+     * @return KeyType,
+     */
+    virtual KeyType GetKeyType() const;
```

Comment:
add override keyword

Reply:
add implementation here. In this case const getters usually automatically inlined AFAIK.

Reply:
not sure virtual functions are inlined well. but anyway I moved implementation into the header.